src/rp2_common/hardware_rtc/rtc.c

```diff
@@ -114,60 +139,88 @@ void rtc_enable_alarm(void) {
     }
 }
 
-static void rtc_irq_handler(void) {
+void rtc_disable_alarm(void) {
+    // Disable matching and wait for it to stop being active
+    hw_clear_bits(&rtc_hw->irq_setup_0, RTC_IRQ_SETUP_0_MATCH_ENA_BITS);
+    while (rtc_hw->irq_setup_0 & RTC_IRQ_SETUP_0_MATCH_ACTIVE_BITS) {
+        tight_loop_contents();
+    }
+}
+
+static void __no_inline_not_in_flash_func(rtc_irq_handler)(void) {
     // Always disable the alarm to clear the current IRQ.
     // Even if it is a repeatable alarm, we don't want it to keep firing.
     // If it matches on a second it can keep firing for that second.
     rtc_disable_alarm();
 
     if (_alarm_repeats) {
-        // If it is a repeatable alarm, re enable the alarm.
-        rtc_enable_alarm();
+        if(_alarm_repeats == CONTINUOUS_REPEAT_EVERY_SEC) {
+          // we need to modify the sec entry with the next valid RTC change and store it into irq_setup_1
+          datetime_t t;
+          rtc_get_datetime(&t);
+          rtc_hw->irq_setup_1 = ADD_AND_ENABLE_REPEATABLE_SECOND(t.sec);
+        }
     }
 
     // Call user callback function
-    if (_callback) {
-        _callback();
+    _callback();
+
+  // If it is a repeatable alarm, re enable the alarm.
+  if(_alarm_repeats) {
+        rtc_enable_alarm();
     }
 }
 
-static bool rtc_alarm_repeats(datetime_t *t) {
-    // If any value is set to -1 then we don't match on that value
-    // hence the alarm will eventually repeat
-    if (t->year  < 0) return true;
-    if (t->month < 0) return true;
-    if (t->day   < 0) return true;
-    if (t->dotw  < 0) return true;
-    if (t->hour  < 0) return true;
-    if (t->min   < 0) return true;
-    if (t->sec   < 0) return true;
-    return false;
+static repeat_type rtc_alarm_repeats(const datetime_t *t) {
+  // If any value is set to -1 then we don't match on that value
+  // hence the alarm will eventually repeat
+    if (t->year < 0 && t->month < 0 && t->day < 0 && t->dotw < 0
+        && t->hour < 0 && t->min < 0 && t->sec < 0) return CONTINUOUS_REPEAT_EVERY_SEC;
```

Comment:
Probably worth adding a comment here saying that if all values are -1 then we repeat every second?